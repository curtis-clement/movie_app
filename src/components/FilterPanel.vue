<script setup lang="ts">
import TextChip from '@/components/TextChip.vue';

interface FilterOption {
  filterName: string;
  options: string[];
  selectedOptions: string[];
}

const props = defineProps<{
  selectedFilterName: string;
  filters: FilterOption[];
}>();

const emits = defineEmits(['toggleFilter', 'chipClick']);

function toggleFilter(filterName: string) {
  emits('toggleFilter', filterName);
}

function handleChipClick(filterName: string, option: string) {
  emits('chipClick', filterName, option);
}

function isOptionSelected(filterName: string, option: string) {
  return props.filters.find(f => f.filterName === filterName)?.selectedOptions.includes(option);
}

function clearFilters() {
  emits('clearFilters');
}
</script>

<template>
  <div class="filter-panel">
    <div class="filter-headers">
      <div
        v-for="filter in props.filters"
        :key="filter.filterName"
        class="filter-header"
        @click="toggleFilter(filter.filterName)"
      >
        {{ filter.filterName }}
        <span class="arrow" :class="{ 'arrow-down': props.selectedFilterName === filter.filterName }">
          â–¼
        </span>
      </div>
      <button class="clear-filters-button" @click="clearFilters">Clear Filters</button>
    </div>

    <div v-if="props.selectedFilterName" class="filter-chips">
      <TextChip
        v-for="option in props.filters.find(f => f.filterName === props.selectedFilterName)?.options"
        :key="option"
        :text="option"
        :selected="isOptionSelected(props.selectedFilterName, option)"
        @emit-click="handleChipClick(props.selectedFilterName, option)"
      />
    </div>
  </div>
</template>

<style scoped>
.filter-panel {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  margin-bottom: 1rem;
}

.filter-headers {
  display: flex;
  gap: .5rem;
}

.filter-header {
  align-items: center;
  border-radius: 4px;
  cursor: pointer;
  display: flex;
  gap: 0.5rem;
  padding: 0.5rem 1rem;
}

.filter-header:hover {
  background-color: #e0e0e0;
}

.arrow {
  display: inline-block;
  font-size: .75rem;
  transform: rotate(-90deg);
  transition: transform 0.2s ease;
}

.arrow-down {
  transform: rotate(0deg);
}

.filter-chips {
  display: flex;
  flex-wrap: wrap;
  gap: 0.5rem;
}

.clear-filters-button {
  background-color: #3498db;
  border: none;
  border-radius: 4px;
  color: white;
  cursor: pointer;
  padding: 0.5rem 1rem;
  transition: background-color 0.2s;
}

.clear-filters-button:hover {
  background-color: #2980b9;
}
</style>