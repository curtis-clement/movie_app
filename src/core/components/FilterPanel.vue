<script setup lang="ts">
import DefaultButton from '@/core/components/DefaultButton.vue';
import TextChip from '@/core/components/TextChip.vue';
import type { FilterOption } from '@/models/filter.model';

const props = defineProps<{
  selectedFilterName: string;
  filters: FilterOption[];
}>();

const emits = defineEmits(['toggle-filter', 'chip-click', 'clear-filters']);

function toggleFilter(filterName: string) {
  emits('toggle-filter', filterName);
}

function handleChipClick(filterName: string, option: string) {
  emits('chip-click', filterName, option);
}

function isOptionSelected(filterName: string, option: string) {
  return props.filters.find(f => f.filterName === filterName)?.selectedOptions.includes(option);
}

function clearFilters() {
  emits('clear-filters');
}
</script>

<template>
  <div class="filter-panel">
    <div class="filter-headers">
      <div
        v-for="filter in props.filters"
        :key="filter.filterName"
        class="filter-header"
        @click="toggleFilter(filter.filterName)"
      >
        {{ filter.filterName }}
        <span class="arrow" :class="{ 'arrow-down': props.selectedFilterName === filter.filterName }">
          â–¼
        </span>
      </div>
      <DefaultButton @button-click="clearFilters">
        <template #text>Clear Filters</template>
      </DefaultButton>
    </div>

    <div v-if="props.selectedFilterName" class="filter-chips">
      <TextChip
        v-for="option in props.filters.find(f => f.filterName === props.selectedFilterName)?.options"
        :key="option"
        :text="option"
        :selected="isOptionSelected(props.selectedFilterName, option)"
        @emit-click="handleChipClick(props.selectedFilterName, option)"
      />
    </div>
  </div>
</template>

<style scoped>
.filter-panel {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  margin-bottom: 1rem;
}

.filter-headers {
  display: flex;
  gap: .5rem;
}

.filter-header {
  align-items: center;
  border-radius: 4px;
  cursor: pointer;
  display: flex;
  gap: 0.5rem;
  padding: 0.5rem 1rem;
}

.filter-header:hover {
  background-color: #e0e0e0;
}

.arrow {
  display: inline-block;
  font-size: .75rem;
  transform: rotate(-90deg);
  transition: transform 0.2s ease;
}

.arrow-down {
  transform: rotate(0deg);
}

.filter-chips {
  display: flex;
  flex-wrap: wrap;
  gap: 0.5rem;
}
</style>