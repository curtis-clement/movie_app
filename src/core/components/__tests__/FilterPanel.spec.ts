import { beforeEach, describe, it, expect } from 'vitest';
import { mount, VueWrapper } from '@vue/test-utils';
import FilterPanel from '@/core/components/FilterPanel.vue';
import type { FilterOption } from '@/models/filter.model';

const props = {
  selectedFilterName: 'Test Filter',
  filters: [
    {
      filterName: 'Test Filter',
      options: ['Option 1', 'Option 2', 'Option 3'],
      selectedOptions: [],
    },
  ],
};

const mountComponent = (props: {
  selectedFilterName: string;
  filters: FilterOption[];
}) => {
  return mount(FilterPanel, {
    props,
  });
};

describe('FilterPanel', () => {
  let wrapper: VueWrapper<InstanceType<typeof FilterPanel>>;

  beforeEach(() => {
    wrapper = mountComponent(props);
  });

  it('renders the component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('passes props correctly', () => {
    expect(wrapper.props()).toEqual(props);
  });

  it('displays the filter name', () => {
    expect(wrapper.text()).toContain('Test Filter');
  });

  it('displays the filter options', () => {
    expect(wrapper.text()).toContain('Option 1');
    expect(wrapper.text()).toContain('Option 2');
    expect(wrapper.text()).toContain('Option 3');
  });

  it('emits toggle-filter event when a filter header is clicked', () => {
    wrapper.find('.filter-header').trigger('click');
    expect(wrapper.emitted('toggle-filter')).toBeDefined();
    expect(wrapper.emitted('toggle-filter')?.[0]).toEqual(['Test Filter']);
  });

  it('emits chip-click event when a filter option is clicked', () => {
    wrapper.find('.text-chip').trigger('click');
    expect(wrapper.emitted('chip-click')).toBeDefined();
    expect(wrapper.emitted('chip-click')?.[0]).toEqual(['Test Filter', 'Option 1']);
  });

  it('emits clear-filters event when the clear filters button is clicked', () => {
    wrapper.find('button').trigger('click');
    expect(wrapper.emitted('clear-filters')).toBeDefined();
  });

  it('calculates isOptionSelected correctly', async () => {
    wrapper.setProps({
      selectedFilterName: 'Test Filter',
      filters: [
        {
          filterName: 'Test Filter',
          options: ['Option 1', 'Option 2', 'Option 3'],
          selectedOptions: ['Option 2', 'Option 3'],
        },
      ],
    });

    await wrapper.vm.$nextTick();
    const textChips = wrapper.findAll('.text-chip');
    expect(textChips[0].classes()).not.toContain('text-chip-selected');
    expect(textChips[1].classes()).toContain('text-chip-selected');
    expect(textChips[2].classes()).toContain('text-chip-selected');
  });
});
