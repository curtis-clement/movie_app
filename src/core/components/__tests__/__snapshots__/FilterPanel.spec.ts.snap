// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilterPanel > renders the component 1`] = `
"<div data-v-b704e385="" class="filter-panel">
  <div data-v-b704e385="" class="filter-headers">
    <div data-v-b704e385="" class="filter-header">Test Filter <span data-v-b704e385="" class="arrow arrow-down"> â–¼ </span></div><button data-v-5f820906="" data-v-b704e385="" class="default-button">
      <div data-v-5f820906="" class="button-content">Clear Filters</div>
    </button>
  </div>
  <div data-v-b704e385="" class="filter-chips">
    <div data-v-be2c2fe7="" data-v-b704e385="" class="text-chip">Option 1</div>
    <div data-v-be2c2fe7="" data-v-b704e385="" class="text-chip">Option 2</div>
    <div data-v-be2c2fe7="" data-v-b704e385="" class="text-chip">Option 3</div>
  </div>
</div>"
`;
